### PR TITLE
stm32 devices with USB-C PD cannot use CC1 and CC2 pins by default

### LINKED ISSUE #33253: STM32G4 with USB-C PD: Some pins cannot be used as input by default
**Describe the bug**

The USB Type-C and Power Delivery Dead Battery feature requires 5k pull-downs on `UCPD1_CC1` and `UCPD1_CC2` pins (`PB4` and `PB6` on STM32G474 series MCUs).

These pull-downs are enabled by default and need to be actively disabled in order to use those pins as inputs.

**To Reproduce**

Configure PB4 or PB6 as inputs using the GPIO API and pull them up externally with e.g. 5k resistor to 3.3V. You will see that they end up in the middle at around 1.65V instead of at 3.3V.

**Expected behavior**

If something is configured as input via the GPIO API it should work out of the box.

These pull-ups seem to be somewhat separate from normal pinmuxing, so I'm not sure where this should be fixed. Possibly in some generic init code for this MCU, as the reference manual says:

> After exiting reset, the USB Type-C “dead battery” behavior is enabled, which may have a pull-down effect on CC1 and CC2 pins. It is recommended to disable it in all cases, either to stop this pull-down or to hand over control to the UCPD1 (which should therefore be initialized before doing the disable)."

**Workaround**

Add following line somewhere in the initialization of the app:

```C
PWR->CR3 |= PWR_CR3_UCPD_DBDIS;
```

### PR DESCRIPTION
**Pull request by FRASTM**
It disables USB Type-C dead battery pull-down behavior on UCPDx_CC1 and UCPDx_CC2 pins
of the stm32G4 and stm32L5  and stm32G0 soc series which have the USB type C Power Delivery feature. 

After exiting reset, the USB Type-C “dead battery” behavior is enabled, which may have a
pulldown effect on CC1 and CC2 pins. It is recommended to disable it in all cases, either to
stop this pull-down or to hand over control to the UCPD (which should therefore be initialized
before doing the disable).

For stm32l5 and stm32g4, the UCPD DBDIS is controlled by the UCPD1_DBDIS bit of the PWR CR3 register.
For stm32g0, the UCPD DBDIS are controlled by the strobe bits of the SYSCFG CFGR1 register

Fix  #33253  

Signed-off-by: Francois Ramu <francois.ramu@st.com>

### CODE REVIEW

File: soc/arm/st_stm32/stm32g4/soc.c [hunk 1]
```diff
@@ -45,6 +49,11 @@ static int stm32g4_init(const struct device *arg)
 	/* allow reflashing board */
 	LL_DBGMCU_EnableDBGSleepMode();
 
+#if defined(PWR_CR3_UCPD_DBDIS)
+	/* Disable USB Type-C dead battery pull-down behavior */
+	__HAL_RCC_PWR_CLK_ENABLE();
```
Comment (erwango) [inline soc/arm/st_stm32/stm32g4/soc.c:54]: Please use LL API

File: soc/arm/st_stm32/stm32l5/soc.c [hunk 2]
```diff
@@ -47,6 +47,11 @@ static int stm32l5_init(const struct device *arg)
 	LL_APB1_GRP1_EnableClock(LL_APB1_GRP1_PERIPH_PWR);
 	LL_PWR_SetRegulVoltageScaling(LL_PWR_REGU_VOLTAGE_SCALE0);
 
+#if defined(PWR_CR3_UCPD_DBDIS)
+	/* Disable USB Type-C dead battery pull-down behavior */
+	__HAL_RCC_PWR_CLK_ENABLE();
```
Comment (erwango) [inline soc/arm/st_stm32/stm32l5/soc.c:52]: LL_APB1_GRP1_EnableClock(LL_APB1_GRP1_PERIPH_PWR); is called just before.

**Review by erwango — CHANGES_REQUESTED** [review]

File: soc/arm/st_stm32/stm32g4/soc.c (hunk 1 above)
Comment (FRASTM) [reply soc/arm/st_stm32/stm32g4/soc.c]: LL api is LL_APB1_GRP1_EnableClock(LL_APB1_GRP1_PERIPH_PWR) from the stm32_ll_bus.h

**Comment by FRASTM** [comment]
with LL API

File: soc/arm/st_stm32/stm32l5/soc.c (hunk 2 above)
Comment (FRASTM) [reply soc/arm/st_stm32/stm32l5/soc.c]: removed

**Comment by erwango** [comment]
^^ @str4t0m @molchec

**Comment by FRASTM** [comment]
including SYSCFG clock control

File: soc/arm/st_stm32/stm32g4/soc.c [hunk 3]
```diff
@@ -15,6 +15,11 @@
 #include <arch/cpu.h>
 #include <arch/arm/aarch32/cortex_m/cmsis.h>
 
+#if defined(PWR_CR3_UCPD_DBDIS)
```
Comment (erwango) [inline soc/arm/st_stm32/stm32g4/soc.c:18]: `PWR_CR3_UCPD_DBDIS` is defined for all G4 variants.
In that regard, this condition is superfluous.

File: soc/arm/st_stm32/stm32g4/soc.c [hunk 4]
```diff
@@ -45,6 +50,11 @@ static int stm32g4_init(const struct device *arg)
 	/* allow reflashing board */
 	LL_DBGMCU_EnableDBGSleepMode();
 
+#if defined(PWR_CR3_UCPD_DBDIS)
```
Comment (erwango) [inline soc/arm/st_stm32/stm32g4/soc.c:53]: ditto

File: soc/arm/st_stm32/stm32l5/soc.c [hunk 5]
```diff
@@ -47,6 +47,10 @@ static int stm32l5_init(const struct device *arg)
 	LL_APB1_GRP1_EnableClock(LL_APB1_GRP1_PERIPH_PWR);
 	LL_PWR_SetRegulVoltageScaling(LL_PWR_REGU_VOLTAGE_SCALE0);
 
+#if defined(PWR_CR3_UCPD_DBDIS)
```
Comment (erwango) [inline soc/arm/st_stm32/stm32l5/soc.c:50]: ditto

**Review by erwango — CHANGES_REQUESTED** [review]

File: soc/arm/st_stm32/stm32g4/soc.c (hunk 4 above)
Comment (FRASTM) [reply soc/arm/st_stm32/stm32g4/soc.c:53]: yes, 
however, the `stm32g4xx_ll_pwr.h` has this flag  `#if defined(PWR_CR3_UCPD_DBDIS)`

File: soc/arm/st_stm32/stm32l5/soc.c (hunk 5 above)
Comment (FRASTM) [reply soc/arm/st_stm32/stm32l5/soc.c:50]: removed

Comment (erwango) [reply soc/arm/st_stm32/stm32l5/soc.c:50]: still there...

**Comment by FRASTM** [comment]
removing the PWR_CR3_UCPD_DBDIS  flag for the stm32L5 serie

**Review by erwango — APPROVED** [review]

**Review by martinjaeger — APPROVED** [review]
Looks good. Thanks for implementing the fix.

**Pull request merged**